### ISSUE TITLE
Environment setup is extremely confusing.

### ISSUE THREAD

**Issue by Garbee** [opulencephp/Opulence#7]
The entire environment setup is quite confusing to use. While it is *possible* to integrate it with other tools such as [phinx](https://phinx.org/), it isn't easy nor convenient.

[Dotenv](https://github.com/vlucas/phpdotenv) is a **fantastic** package maintained by Vance that provides a seamless environment configuration experience. Simple to use and share between tooling.

Could a move to using Dotenv be considered? Taking advantage of it will provide the best experience to developers.

**Comment by OddGreg** [comment]
I would like to second Garbee's suggestion. I've already added `Dotenv` to the sample app (among other things) and am working on integrating it.

**Comment by opulencephp** [comment]
Can I ask what specifically is confusing about the environment library in Opulence?  What does Dotenv provide that Opulence's library doesn't?  That's not a loaded question - I really want to know.  I'm not a fan of the fact that Dotenv's config files are not PHP files, which forces you to learn their custom syntax to do something that is easily achievable in PHP.  Why lose IDE autocompletion support?

**Comment by OddGreg** [comment]
Firstly, I don't think it is particularly confusing at all. However, many of us are used to using a `.env` file (usually in the root) that describes a number of environment settings. I use this extensively and, admittedly, it is comfortable to use. (DB setting, debug settings and etc.)

**Comment by Garbee** [comment]
>   I'm not a fan of the fact that Dotenv's config files are not PHP files, which forces you to learn their custom syntax to do something that is easily achievable in PHP.

It is true they aren't PHP files, sure. However they are extremely simple and a well known config type for ages. It is the [INI format](https://en.wikipedia.org/wiki/INI_file). So they are far from "custom syntax".

> Can I ask what specifically is confusing about the environment library in Opulence?

Registering the different formats of auto-detecting environments for one. It is first-in-first-out and not weighted. So you need to order things just right and have them as specific as possible to get it working. Instead of just having a single file that says "your env is this."

Also, just having different files for each env. In an ideal system, this doesn't matter. You have *one* file to explain to the system the variables, that's it. No managing multiple (mostly redundant) options.

Further, you have all the great features for free, like requiring certain variables be provided. No extra maintenance within Opulence to manage that.

> Why lose IDE autocompletion support?

You don't have it in the first place. All you have is "X obj has Y method that takes *some string* for the identifier." Same thing as `$env["DB_HOST"]` or any other shim you'd like to make. Claiming you have some kind of "autocompletion support" here is a facade since it doesn't truly exist. You get no valuable assistance from the current environment setup compared to using Dotenv.

**Comment by OddGreg** [comment]
Ok, so I fiddled around with it for a while and can see where this is simply a design choice. Scanning or transferring configuration options from whatever `.env~` is active depending on start up conditions is trivial. It would be nice to handle it differently, but I don't know if you really need to change it. 

I've experimented with several options and find that I can work without the need to change anything. It's a design choice that isn't a deal-breaker in my view. And in any case, adding `Dotenv` and using it is trivial.

**Comment by Garbee** [comment]
It isn't "necessary". However, this is something that can be made much simpler and straightforward for new users.

**Comment by OddGreg** [comment]
Hard to say. Certainly to seasoned developers, yes. However, adding it and using it in the application  or an extending package is as simple as:
```
/**
 * ----------------------------------------------------------
 * Load the environment.
 * ----------------------------------------------------------
 */
$dotenv = new Dotenv\Dotenv(__DIR__ . '/../../');
$dotenv->load();
```
... somewhere such as added to the index.php file, and replacing ...

`$environment->getName() == Environment::PRODUCTION,` 
with  `$environment->getName() == getenv('ENVIRONMENT'),`

assuming that `.env` includes `ENVIRONMENT = development` It's not a _breaking_ design choice.

Any other setting may be handled similarly, while the application only depends on the `getenv()`.

What I mean is that it isn't a deal breaker at all, and alternative methods aren't difficult. Even if only a mention in the documentation gives an example. I'm not in disagreement with you on the salient points, but I don't think that he NEEDS to change anything for the framework to be useful to anyone who wants to extend it or develop an application with it.

**Comment by opulencephp** [comment]
I've removed the auto-detection of your current environment based on host name matching from the [skeleton project](https://github.com/opulencephp/Project).  Now, it [uses the value in "ENV_NAME" to set the environment name](https://github.com/opulencephp/Project/blob/1.0/config/environment.php).  However, I did leave the environment resolver in the Environment library because some people may find it useful.  I am going to close this issue and add config validation as a feature request.
